R/tags.R

```diff
@@ -687,18 +733,40 @@ findDependencies <- function(tags, tagify = TRUE) {
   if (isTRUE(tagify)) {
     tags <- tagify(tags)
   }
-  dep <- htmlDependencies(tags)
-  if (!is.null(dep) && inherits(dep, "html_dependency"))
-    dep <- list(dep)
+  deps <- resolveFunctionalDependencies(htmlDependencies(tags))
   children <- if (is.list(tags)) {
     if (isTag(tags)) {
       tags$children
     } else {
       tags
     }
   }
-  childDeps <- unlist(lapply(children, findDependencies, tagify = FALSE), recursive = FALSE)
-  c(childDeps, if (!is.null(dep)) dep)
+  childDeps <- unlist(lapply(children, findDependencies, tagify = FALSE), recursive = FALSE, use.names = FALSE)
+  c(childDeps, deps)
+}
+
+#' Resolve tag functions
+#'
+#' Resolves any [tagFunction()]s inside a list of [htmlDependencies()]. To
+#' resolve [tagFunction()]s _and then_ remove redundant dependencies all at once,
+#' use [resolveDependencies()] (which calls this function internally).
+#'
+#' @inheritParams resolveDependencies
+#' @export
+#' @return a list of HTML dependencies.
+#' @seealso [tagFunction()], [resolveDependencies()]
+resolveFunctionalDependencies <- function(dependencies) {
+  dependencies <- asDependencies(dependencies)
+  dependencies <- lapply(dependencies, function(dep) {
+    if (is_tag_function(dep)) {
+      dep <- dep()
+    }
+    if (isTag(dep) || isTagList(dep)) {
```

Comment:
Do we need this `if` code path? I thought that that we could assume that whatever is going into this function is a list of `htmlDependency`s and/or functions returning `htmlDependency`s -- and that no tag objects were allowed.

Reply:
I just remembered that (after our discussion), that this case is actually very important because of

```r
#' @export
as.tags.html_dependency <- function(x, ...) {
  attachDependencies(tagList(), x)
}
```

Reply:
A comment here explaining this would help future us (or whoever's working on this code).

Also, if our tests don't already cover this, we should add ones that do.

Reply:
Actually, I don’t think we need the case at all since we should always be receiving an object that has passed through `htmlDependencies(x)`, so I added a warning in https://github.com/rstudio/htmltools/pull/180/commits/d9ca877a73cd175de3371c75f375199979818dff